### PR TITLE
Kazantsev Evgeny Lab2. var3

### PR DESCRIPTION
**Pull request by neotouch1**
This pass can be used to automatically inject custom functions at the beginning and end of each detected loop in LLVM-IR code.

### CODE REVIEW

File: llvm/labs/lab2/kazantsev_evgeny/FindLoopPlug.cpp [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+    auto LoopEndFunction = F.getParent()->getOrInsertFunction("loop_end", FM);
+    IRBuilder<> Builder(F.getContext());
+
+    auto &LA = FA.getResult<LoopAnalysis>(F); // Gets the result of loop analysis
+
+    for (auto &L : LA) {
+      auto *Head = L->getHeader();
+      for (auto *const Pred : children<Inverse<BasicBlock *>>(Head)) {
+        if (!L->contains(Pred)) {
+          Builder.SetInsertPoint(Pred->getTerminator());
```
Comment (m-ly4) [inline llvm/labs/lab2/kazantsev_evgeny/FindLoopPlug.cpp]: Please insert loop_start/loop_end calls only if they are not present already.
And add a test for that case.

**Review by m-ly4 — CHANGES_REQUESTED** [review]
Please make more descriptive MR title

File: llvm/labs/lab2/kazantsev_evgeny/FindLoopPlug.cpp [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+          }
+          if (!loopStartPresent) {
+            Builder.SetInsertPoint(Pred->getTerminator());
+            Builder.CreateCall(LoopStartFunction); // creates a function call
+          }
+        }
+      }
+
+      SmallVector<BasicBlock *, 8> ExitB;
+      L->getExitBlocks(ExitB);
```
Comment (m-ly4) [inline llvm/labs/lab2/kazantsev_evgeny/FindLoopPlug.cpp]: To avoid many times processing of one basic block you can use method `getUniqueExitBlocks`.

**Review by m-ly4 — CHANGES_REQUESTED** [review]
Please fill PR title in accordance with other PRs.

File: llvm/labs/lab2/kazantsev_evgeny/FindLoopPlug.cpp (hunk 2 above)
Comment (neotouch1) [reply llvm/labs/lab2/kazantsev_evgeny/FindLoopPlug.cpp]: used it

**Comment by neotouch1** [comment]
> Please fill PR title in accordance with other PRs.

I did

**Review by m-ly4 — APPROVED** [review]

**Pull request merged**